src/wallet/mms.cpp

```diff
@@ -0,0 +1,825 @@
+	  data.processing = message_processing_sign_tx;
+	  data.message_ids.push_back(m.id);
+	  data_list.push_back(data);
+	  return true;
+	}
+      }
+    }
+  }
+  if (waiting_found)
+  {
+    wait_reason = tr("Waiting message is not a tx and thus not processable now.");
+  }
+  else
+  {
+    wait_reason = tr("There is no message waiting to be processed.");
+  }
+
+  return false;
+}
+
+void message_store::set_messages_processed(const processing_data &data)
+{
+  for (size_t i = 0; i < data.message_ids.size(); ++i)
+  {
+    set_message_processed_or_sent(data.message_ids[i]);
+  }
+}
+
+void message_store::set_message_processed_or_sent(uint32_t id)
+{
+  uint32_t index = get_message_index_by_id(id);
+  message &m = m_messages[index];
+  if (m.state == message_state_waiting)
+  {
+    m.state = message_state_processed;
+    // So far a fairly cautious and conservative strategy: Only delete from Bitmessage
+    // when fully processed (and e.g. not already after reception and writing into
+    // the message store file)
+    delete_transport_message(id);
+  }
+  else if (m.state == message_state_ready_to_send)
+  {
+    m.state = message_state_sent;
+  }
+}
+
+void message_store::encrypt(uint32_t member_index, const std::string &plaintext, 
+			    std::string &ciphertext, crypto::public_key &encryption_public_key, crypto::chacha_iv &iv)
+{
+  crypto::secret_key encryption_secret_key;
+  crypto::generate_keys(encryption_public_key, encryption_secret_key);
+  
+  crypto::key_derivation derivation;
+  crypto::public_key dest_view_public_key = m_members[member_index].monero_address.m_view_public_key;
+  crypto::generate_key_derivation(dest_view_public_key, encryption_secret_key, derivation);
+
+  crypto::chacha_key chacha_key;
+  crypto::generate_chacha_key(&derivation, sizeof(derivation), chacha_key);
+  iv = crypto::rand<crypto::chacha_iv>();
+  ciphertext.resize(plaintext.size());
+  crypto::chacha20(plaintext.data(), plaintext.size(), chacha_key, iv, &ciphertext[0]);
+}
+
+void message_store::decrypt(const std::string &ciphertext, const crypto::public_key &encryption_public_key, const crypto::chacha_iv &iv,
+			    std::string &plaintext)
+{
+  wallet_state state;
+  m_wallet_state_provider->get_wallet_state(state);
+  crypto::secret_key& view_secret_key = state.view_secret_key;
+  if (state.multisig)
+  {
+    if (state.original_keys_available)
+    {
+      view_secret_key = state.original_view_secret_key;
+    }
+    else
+    {
+      throw std::invalid_argument("Own view secret key not available");
+    }
+  }
+  crypto::key_derivation derivation;
+  crypto::generate_key_derivation(encryption_public_key, view_secret_key, derivation);
+  crypto::chacha_key chacha_key;
+  crypto::generate_chacha_key(&derivation, sizeof(derivation), chacha_key);
+  plaintext.resize(ciphertext.size());
+  crypto::chacha20(ciphertext.data(), ciphertext.size(), chacha_key, iv, &plaintext[0]);
+}
+
+void message_store::send_message(uint32_t id)
```

Comment:
Well, "higher up" i.e. in simplewallet both cases are supported: Not immediately sending a message that was generated, and on the other hand re-sending an existing message because e.g. it was not transported properly. For those use cases separate "add" and "send" methods are needed.